### ISSUE TITLE
allow special characters in Album and Song titles and sub/secondary titles

### ISSUE THREAD

**Issue by waynedpj** [indiehd/web-api#125]
ahoy all,

  first glad to see that indieTorrent .. er i mean indieHD is alive and well :)  i have been following since the torrent days (i miss those!) and happy to have some music to publish.  while we like to have our music available for pay-what-you-wish, we are thrilled to support indieHD since it is all FLOSS, thanks Ben & co!

  our album has a subtitle as well as song titles with various special characters like `(,),?,:,!,`.  it looks like the secondary titles are already put into parentheses for display which handles the album's subtitle in our case.  however we have song titles like "happy? part 4. the Realization station" which are not being accepted currently.

  while i am sure that there are security, etc. issues that i am not aware of, it seems restrictive to deny almost all special characters, since with HTML escaping, etc. there must be a way to allow them safely as other sites do?

  i did a little research and it seems like the FormRequest is where the validation occurs, though currently i cannot yet find the actual validation code?  hopefully it is just a simple change to some Validator?  of course i have never used Laravel nor done major PHP so take this with a large dose of salt!  though if it *is* a "simple" change, perhaps you can give me a nudge in the right direction and i can take a look myself?

  regardless, thanks again for the great FLOSS service and keep up the good work!

peace, w

**Comment by cbj4074** [comment]
Great to hear from you, @waynedpj ! We appreciate your support over the last decade, and hope to have it through the next!

Before we dive into semantics, I should mention that the FLOSS version of the service isn't in production yet, which is why you can't find the actual validation logic. We're referring to what's currently at indiehd.com as the "legacy site" as we work to build-out the new service, which is entirely API-driven.

That said, we can certainly modify or fix issues in the legacy code between now and when we push the new code-base into production. We will need to address this in the new code-base, too, so it should be straightforward to apply the same logic to the legacy site.

With regard to allowing special characters in artist, album, and track names, there's no problem with displaying the characters in HTML; the problem is when we try to download files with with, for example, a `?` in the name, which fails on some platforms, such as Windows. Bear in mind, too, that the download may be in the form of a ZIP file, so the problem doesn't occur until the user attempts to unzip the file, at which time the individual tracks are extracted to the filesystem.

The obvious solution to this is to remove or otherwise escape those characters only when downloading files. To do this, we'd need a comprehensive list of which characters are illegal in filenames across the filesystems used commonly in all major computing platforms (Mac, Windows, and GNU/Linux). This thread provides a decent head-start:

https://stackoverflow.com/questions/1976007/what-characters-are-forbidden-in-windows-and-linux-directory-names

The solution may end-up being slightly more nuanced, but that is the crux of the issue.

I'll spend some time on this over the next couple of days and see what I can come up with...

**Comment by waynedpj** [comment]
> Great to hear from you, @waynedpj ! We appreciate your support over the last decade, and hope to have it through the next!

100% ditto @cbj4074 !


> Before we dive into semantics, I should mention that the FLOSS version of the service isn't in production yet, which is why you can't find the actual validation logic. We're referring to what's currently at indiehd.com as the "legacy site" as we work to build-out the new service, which is entirely API-driven.

well that explains a lot ;)  i thought that it was odd that i could not find any validation code!



> The obvious solution to this is to remove or otherwise escape those characters only when downloading files. To do this, we'd need a comprehensive list of which characters are illegal in filenames across the filesystems used commonly in all major computing platforms (Mac, Windows, and GNU/Linux). This thread provides a decent head-start:
> 
> https://stackoverflow.com/questions/1976007/what-characters-are-forbidden-in-windows-and-linux-directory-names

thanks and let me know what you find.  i would hope that there is a shared sanitize library code available to reuse, e.g. https://stackoverflow.com/questions/2668854/sanitizing-strings-to-make-them-url-and-filename-safe assuming the legacy code is PHP.  among the more promising recommendations IMHO are [`sanitize_file_name` from WordPress](https://codex.wordpress.org/Function_Reference/sanitize_file_name) and [URLify for PHP](https://github.com/jbroadway/urlify).

thanks again and excited to *finally* share some music via indieTorrent :)

**Comment by cbj4074** [comment]
Apologies for the delayed reply @waynedpj !

Thanks for the links and ideas; very helpful! The StackOverflow link is jam-packed with useful responses.

At this point, I've spent quite a bit of time researching this issue, and the more I read, the more I realize just how complex it is (and why I've been kicking this can down the road for so long!).

Fortunately, it looks like there are some viable options in the way of existing third-party libraries.

For the moment, I'd like to focus only on the filename piece, as I don't believe that we use Artist or Album names in any URLs on the legacy site.

I don't love the WordPress option because it's not entirely portable. It's cumbersome to bring just the required code into the project, and `sanitize_file_name()` requires the `apply_filters()` function, too.

Beyond that, I'd prefer to use a well-tested library that has been exposed to some scrutiny, such as https://github.com/OWASP/PHP-ESAPI .

https://github.com/alixaxel/phunction looks promising, too. Judging by his contributors to the StackOverflow thread, the author seems to be fairly well-versed in the subject matter at hand.

I'm going to pull down both of the above libraries and see if either or both is capable of addressing the immediate need.

I'll follow-up here as soon as I'm able. Thanks for your patience!

**Comment by waynedpj** [comment]
@cbj4074 thank you for the update and your time!  i figured it would be a bigger issue than i had hoped .. user "feature" requests are almost always like that ;)  i had seen phunction but since the GH repo had not been updated in 5+ years it fell off my list.  but hopefully the code can be helpful?

either way you are the expert!  let me know if i can help besides making more music to share ;)  thanks as always :)

**Comment by cbj4074** [comment]
@waynedpj Good call! I didn't notice how stale *both* libraries (phunction and PHP-ESAPI) are at first. Neither one inspires much confidence, given how they've languished.

After re-reading virtually all of the comments on https://stackoverflow.com/questions/2668854/sanitizing-strings-to-make-them-url-and-filename-safe , one library you mentioned initially stood-out: [URLify for PHP](https://github.com/jbroadway/urlify); it seems perfect for what we need in this context.

In particular, the `URLify::downcode()` method by itself performs the bulk of the sanitation that is necessary for our use-case.

After playing with the library a bit, I think it will work. There are some small issues with the code that will necessitate a merge-request to address, but I'll submit one as soon as I'm able; hopefully, it's considered without much delay.

I'm writing tests for our portion now, and hope to finish those today. Once we're sure it does what we need, I'll investigate applying it to the legacy code-base.

Another update soon...

**Comment by cbj4074** [comment]
Upon further consideration, it may not be beneficial to use `URLify::downcode()` here, primarily because doing so would strip a great deal of perfectly acceptable Unicode characters from filenames.

Historically, we haven't accepted *any* characters outside of `[a-zA-Z]` (and a handful of symbols), which is a major limitation to the naming convention. This lack of support was rooted mostly in the fact that UTF-8 was poorly, if at all, supported in some of the implementations with which our code-base integrates.

For example, [ID3v2.3 doesn't support UTF-8 at all](https://stackoverflow.com/a/9861839), and it seemed simplest just to use the same validation scheme on all strings that could conceivably end-up in an IDv3-tagged file within our catalog.

As we've established in this thread and those that were precursors to it, there exists a fundamental distinction between text that is merely displayed to the user-agent and text that is encoded into downloaded audio files or used in their names, and we should make this distinction in code (including in the legacy code-base).

At the very minimum, we should be able to display all printable Unicode characters in the browser, as doing so has no bearing on audio tags or filenames.

But before we can basically remove all restrictions on Artist, Album, and Song names, we need to implement a method for sanitizing those strings against a small subset of characters anytime the following actions are performed within the framework:

1. Embedding an ID3v2.x tag into an MP3 file.
    a. Must ensure that we're using ID3v2.4 or later.
        i. This process involves validation code that requires [getID3](http://getid3.sourceforge.net/), which, in turn, means that getID3 must support ID3v2.4 or later (I just confirmed that it does).
2. Embedding a metadata tag into other supported audio formats (currently, FLAC, ALAC, and AAC/MP4).
    a. It seems likely that these other formats have supported UTF-8 for quite some, but that must be confirmed.
3. Presenting individual files to the user-agent for download.
4. Packing ZIP archives containing multiple files (and a parent directory) for download (affects the file and directory names *within* the archive, as well as the archive name itself).

I may be missing others, but those are certainly the most prominent scenarios that come to mind without stepping through the actual code.

**Comment by cbj4074** [comment]
To address the immediate sanitization need, I cooked-up a simple library that we can pull into both the legacy code-base and the in-progress API, because I couldn't find anything that meets our needs exactly without being overly heavy-handed:

https://github.com/indiehd/filename-sanitizer

In the event that we encounter an integration surface that doesn't support UTF-8, we can utilize the URLify library to "downcode" into the reproducible character range.

At this point, I need to add the new library to the legacy project, comb through the code-base, implement the filter, and see what breaks. Unfortunately, that code is devoid of tests, so it'll be somewhat of a manual process.

I'll reply here regarding any progress!

**Comment by waynedpj** [comment]
@cbj4074 thanks for the update and work!  so basically we need to sanitize only user provided strings that are going to end up as filenames, for the actual track or artwork files and archive files which contains these files, as all the metadata container formats and user agents support UTF-8?

if there is a way that i can do some manual testing on the legacy code i can give it a shot.

thanks again!

**Comment by cbj4074** [comment]
@waynedpj Yes, exactly.

I confirmed that FLAC and MP4 (M4A) both support UTF-8, so we should be in good shape in that respect.

We use FLAC's own tool to write tags to FLAC files, so that's a non-issue. 

We use [AtomicParsley](http://atomicparsley.sourceforge.net/) for writing the M4A tags, which appears to deal in UTF-8, internally, so that's a non-issue, too.

Do you speak/write in any language other than English, by chance?

**Comment by waynedpj** [comment]
> Do you speak/write in any language other than English, by chance?

mother tongue is English but i do speak/read/write Italian decently and un muy muy poco spangolo.

**Comment by cbj4074** [comment]
@waynedpj That's probably sufficient for our purposes here. :)

Even though you can't title your album/songs precisely as you'd like just yet, have you uploaded your content to the site already (using "placeholder" names)? If not, you may want to go ahead and do that at your convenience, as I think it will be helpful to have your data available as we progress in this effort.

And beyond that, it would be excellent for you to lay eyes on the overall workflow, as I'm sure you'll have plenty of other feedback along the way.

**Comment by waynedpj** [comment]
@cbj4074 OK i uploaded the album with placeholder/modified titles to get passed the validation https://indiehd.com/catalog/songs/album/1968/

thanks again!

PS: why is that the album URL instead of something like https://indiehd.com/ingiro/releases/demonstrate/ ?  even the indieHD URL of https://indiehd.com/ingiro gets redirected to https://indiehd.com/catalog/albums/artist/1123/page/1/ ?

**Comment by cbj4074** [comment]
@waynedpj Nice work!

The short answer to your question regarding the redirection behavior is rooted in URL canonicalization and application caching.

On the new site (the one we're building now, on GitHub), the URLs will be much "friendlier".

I'm working on my piece of this right now and should have another update later today.

**Comment by cbj4074** [comment]
While rooting-around in the source-code, I notice that I made the following annotation in the MP3 tagging function:

```
// Certain applications cannot read UTF-8 tags,
// such as the Explorer shell in Windows Vista.
```

I mention this because we may find that we need to add a "legacy tags" option that writes non-UTF8 tags on an on-demand, per-download basis.

**Comment by cbj4074** [comment]
Apologies for the delay on this end.

The work on this is nearly complete, but further testing is required before the changes are deployed.

I am hoping that we can move the changes into production by early next week.

I'll let you know if there's anything you can do to facilitate testing @waynedpj !

**Comment by waynedpj** [comment]
thanks for the update and progress!  so maybe this will be ready before our next release .. and its continuation of our troublesome song titles ;)  please let me know if there is anything with which i can help and thanks again!

**Comment by cbj4074** [comment]
I'm pleased (and relieved!) to report that this Feature Request has been implemented and pushed into production.

There's one outstanding issue, which is that file-names that

a) exceed the file name length imposed on the Windows platform (and thus necessitate shortening to some extent), and;
b) contain characters that [jbroadway/urlify](https://github.com/jbroadway/urlify) does not "downcode" by default (that is, without adding custom encodings)

are omitted from full-album (ZIP archive) downloads. The logic in the archiving routine skips these files because their names are too long to be accessed reliably on the Windows platform.

That said, there are exactly zero albums in the catalog currently that are affected, so this limitation is very much an "edge-case".

At your convenience, please go ahead and edit your exiting album @waynedpj , and when you're ready, upload the new one, and name it however you like.

To be clear, the parenthetical behavior for album and song names has been retained (i.e., the "alternate track name"), *but*, you can ignore it in favor of using whichever special characters you like in the primary name. In other words, you should now be able to jam whichever characters you like in the primary song title/name, and can omit the parenthetical (alternate) name entirely.

Please let us know how this works for you, whenever you have a chance to fiddle with it.

Thanks again for your continued interest and patience!

**Comment by waynedpj** [comment]
thank you @cbj4074 !!  wow, i remembering emailing you how many years ago with this request .. better late than never :)  thanks so much!!

i edited the track and album data and it all seems to be working save 1 small issue:
- 2 tracks (6 & 8) use an apostrophe `'` in their titles which is changed in the downloaded file name to `&#39;`.  after some research i [discovered](https://en.wikipedia.org/wiki/Apostrophe#Entering_apostrophes) that i was using a 'typewriter apostrophe' (who knew?!).  i then tried changing it to a 'typographic apostrophe' and everything looked normal.  while i have read that typographic apostrophes are preferred to the typewriter version these days, should this new filter work with the classic typewriter apostrophe as well?  or at least convert all typewriter apostrophes to typographic ones?

the 2 track filenames with the typewriter apostrophe end up as:
06 happy part 3. you&#39;ve got some explaining To Do!.mp3
08 to Lose one&#39;s self Completely (i. new seeds ii. rhapsodize).mp3

i see this in both my terminal as well as the file manager and file dialogs on my GNU+Linux OS.

again, not a big issue but i wanted to point it out in case you want to address it.

regardless, thanks again!!!!  now i can make the next album's titles as wacky as i care to ;)

**Comment by waynedpj** [comment]
PS: i left track 8 with the typewriter apostrophe for now in case you wanted to take a look; track 6 uses the typographic apostrophe.

**Comment by cbj4074** [comment]
No problem @waynedpj , I'm glad to have implemented this long-standing feature request.

I'm pleased to hear that everything seems to be working for your use-case, even if with a few small exceptions.

Yes, very good, please do leave track 8 with the typewriter apostrophe in place so I can take a look.

> should this new filter work with the classic typewriter apostrophe as well? or at least convert all typewriter apostrophes to typographic ones?

It's difficult to say until I dig into it. :) But I'll post back with more information and/or progress soon.

Thanks again for all your help and persistence with this!

**Comment by waynedpj** [comment]
> Thanks again for all your help and persistence with this!

@cbj4074 no thank you!!  and let me know if you need any testing help with this little issue.

**Comment by cbj4074** [comment]
@waynedpj We've created a new issue (noted just above this comment) to track the anomaly you discovered regarding apostrophes, so be sure to follow that issue if you'd like notifications.

**Comment by waynedpj** [comment]
@cbj4074 subscribed, thanks for the update and work!
